### PR TITLE
#9755: Centralize concat implementation to ttnn and delete/replace the old ttlib version from the codebase

### PR DESCRIPTION
**Pull request by sjameelTT**
### Ticket
#9755 

### Problem description
Concat has a divided implementation right now.

### What's changed
-Replaced C++ usages
-Replaced Python usages
-Deleted ttl implementation

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+#include <ranges>
+
+
+namespace ttnn {
+namespace operations {
+namespace data_movement {
+
+    // Wrapper for TTDNN
+    ttnn::Tensor ConcatOperation::operator()(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:19]: move to cpp

Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:19]: fix identation

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+#include "ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.hpp"
+#include "ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp"
+
+#include <ranges>
+
+
+namespace ttnn {
+namespace operations {
+namespace data_movement {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:16]:
```suggestion
namespace ttnn::operations::data_movement {
```

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        int dim,
+        const std::optional<MemoryConfig>& memory_config,
+        std::optional<ttnn::Tensor> optional_output_tensor) {
+        constexpr uint8_t DefaultQueueId = 0;
+        return operator()(DefaultQueueId, input_tensors, dim, memory_config, optional_output_tensor);
+    }
+};
+
+}
+}
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:113]:
```suggestion
}
```

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp [hunk 4]
```diff
@@ -7,7 +7,7 @@
 #include "ttnn/tensor/types.hpp"
 #include "ttnn/operations/core/core.hpp"
 
-#include "ttnn/deprecated/tt_dnn/op_library/concat/concat_op.hpp"
+#include "ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp:10]: I think you really only need "ttnn/decorators.hpp"

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.hpp [hunk 5]
```diff
@@ -7,13 +7,11 @@
...
 
-namespace tt {
-
-namespace tt_metal {
+namespace ttnn::operations::data_movement {
 
 enum class ConcatOpParallelizationStrategy { MULTI_CORE, SHARDED_MULTI_CORE };
 
-struct Concat {
+struct ConcatDeviceOperation {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.hpp:14]: Going forward if you do larger changes like this, follow the latest TMP infra from Akhmed. See Example operation for.. an example

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by skhorasganiTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp (hunk 1 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:19]: Done

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp (hunk 2 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:16]: Done

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp (hunk 3 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:113]: Done

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp (hunk 4 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp:10]: Done

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.hpp (hunk 5 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.hpp:14]: I'll do the next PR with the new infra and eventually refactor everything into the new infra

**Review by tt-rkim — APPROVED** [review]

**Comment by sjameelTT** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10186248527 post commit
https://github.com/tenstorrent/tt-metal/actions/runs/10169288008 T300 model perf
https://github.com/tenstorrent/tt-metal/actions/runs/10185034624/job/28174669296 Frequent tests (fail here matches main)
https://github.com/tenstorrent/tt-metal/actions/runs/10169299987 Model perf
https://github.com/tenstorrent/tt-metal/actions/runs/10185030815 device perf (fail here matches main)

**Review by djordje-tt — APPROVED** [review]

**Review by ntarafdar — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Pull request merged**